### ISSUE TITLE
Thunder's horizontal fusion is memory inefficient for backward functions with activation checkpointing

### ISSUE THREAD

**Issue by IvanYashchuk** [Lightning-AI/lightning-thunder#1337]
TL;DR: Thunder's fusion pass needs to change to consider the memory usage of the operations and the intermediate tensors. It should avoid fusing operations that increase peak memory usage. Use `memory_peak_efficient_func` as the target function for optimization.

Let's take a look at the following code snippet:

```python
import torch
import gc

def memory_peak_efficient_func(t0s, a):
    for t0 in t0s:
        t1 = torch.nn.functional.relu(t0); del t0
        t2 = torch.matmul(t1, t1); del t1
        t3 = torch.nn.functional.relu(t2); del t2
        a = torch.matmul(t3, a); del t3
    return a

N_PARALLEL_PATHS = 10
t0s = [torch.randn(256, 256, device="cuda") for _ in range(N_PARALLEL_PATHS)] # 0.25 MiB * N_PARALLEL_PATHS
a = torch.randn(256, 256, device="cuda") # 0.25 MiB

# Record peak memory usage
gc.collect(0)
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats()
before_in_MB = torch.cuda.max_memory_allocated() / (1024 * 1024)
memory_peak_efficient_func(t0s, a)
max_allocated_in_MB = torch.cuda.max_memory_allocated() / (1024 * 1024) - before_in_MB
print(f"Peak memory usage diff: {max_allocated_in_MB:.2f} MiB")
```
```plaintext
Peak memory usage diff: 0.75 MiB
```
The code snippet above is a simple example of a memory-efficient function that processes a list of tensors sequentially. The function avoids unnecessary memory allocations due to the order of operations and deleting intermediate tensors after they are no longer needed.

The code can be further "optimized" by fusing some operations into a single region. Before fusion let's take a look at the differently structured code snippet below:

```python
import torch
import gc

def memory_peak_inefficient_func(t0s, a):
    # Can be fused into a single region
    t1s = []
    for t0 in t0s:
        t1 = torch.nn.functional.relu(t0); del t0
        t1s.append(t1)
    del t0s

    t2s = []
    while t1s:
        t1 = t1s.pop()
        t2 = torch.matmul(t1, t1); del t1
        t2s.append(t2)
    del t1s

    # Can be fused into a single region
    t3s = []
    while t2s:
        t2 = t2s.pop()
        t3 = torch.nn.functional.relu(t2); del t2
        t3s.append(t3)
    del t2s

    while t3s:
        t3 = t3s.pop()
        a = torch.matmul(t3, a); del t3
    del t3s

    return a

N_PARALLEL_PATHS = 10
t0s = [torch.randn(256, 256, device="cuda") for _ in range(N_PARALLEL_PATHS)] # 0.25 MiB * N_PARALLEL_PATHS
a = torch.randn(256, 256, device="cuda") # 0.25 MiB

# Record peak memory usage
gc.collect(0)
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats()
before_in_MB = torch.cuda.max_memory_allocated() / (1024 * 1024)
memory_peak_inefficient_func(t0s, a)
max_allocated_in_MB = torch.cuda.max_memory_allocated() / (1024 * 1024) - before_in_MB
print(f"Peak memory usage diff: {max_allocated_in_MB:.2f} MiB")
```
```plaintext
Peak memory usage diff: 2.75 MiB
```

The code snippet above is a modified version of the previous code snippet computing the same result. It precomputes intermediate tensors and stores them in lists to be used later. This version of the code is less memory-efficient because it stores intermediate tensors in memory, which increases peak memory usage.

Let's apply Thunder on the inefficient code snippet to see if it can optimize the memory usage.

```python
import thunder

jit_memory_peak_inefficient_func = thunder.jit(memory_peak_inefficient_func)
gc.collect(0)
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats()
before_in_MB = torch.cuda.max_memory_allocated() / (1024 * 1024)
jit_memory_peak_inefficient_func(t0s, a)
max_allocated_in_MB = torch.cuda.max_memory_allocated() / (1024 * 1024) - before_in_MB
print(f"Peak memory usage diff: {max_allocated_in_MB:.2f} MiB")
```
```plaintext
Peak memory usage diff: 2.25 MiB
```

Thunder was able to optimize the memory usage of the inefficient code snippet by fusing the operations into a single region. The peak memory usage decreased from 2.75 MiB to 2.25 MiB, but it is still higher than the memory-efficient version of the code.

What would happen if we apply Thunder on the memory-efficient code snippet?

```python
import thunder

jit_memory_peak_efficient_func = thunder.jit(memory_peak_efficient_func)
gc.collect(0)
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats()
before_in_MB = torch.cuda.max_memory_allocated() / (1024 * 1024)
jit_memory_peak_efficient_func(t0s, a)
max_allocated_in_MB = torch.cuda.max_memory_allocated() / (1024 * 1024) - before_in_MB
print(f"Peak memory usage diff: {max_allocated_in_MB:.2f} MiB")
```
```plaintext
Peak memory usage diff: 2.25 MiB
```

The same memory usage as the inefficient code snippet! Thunder was not able to optimize the memory usage of the memory-efficient code snippet because it applies topological sorting to the computation graph preferring forming as big fusion groups as possible. In this case, the memory-efficient code snippet already has the optimal order of operations, and Thunder breaks it by fusing operations into a single region.

<details>
  <summary>Here's the dataflow graph of the execution trace of the memory-efficient code snippet:</summary>

```py
from thunder.core.transform_common import unwrap_return_value
from thunder.examine import make_trace_dot

t = unwrap_return_value(thunder.last_traces(jit_memory_peak_efficient_func)[-3])
dot = make_trace_dot(t)
```
<img width="730" alt="image" src="https://github.com/user-attachments/assets/c80ccfbd-2c0d-48f6-8652-f0cc44a49474">

</details>

<details>
  <summary>And here's how it looked before the horizontal fusion:</summary>

```py
from thunder.examine import make_trace_dot

t = thunder.last_traces(jit_memory_peak_efficient_func)[0]
dot = make_trace_dot(t)
```
![image](https://github.com/user-attachments/assets/d2b9afbd-7794-4f97-b47e-8e094b26a8d5)

A lot more freedom in terms of grouping and reordering of operations as there are many parallel paths.

</details>

Thunder's fusion pass needs to change to consider the memory usage of the operations and the intermediate tensors. It should avoid fusing operations that increase peak memory usage. Here are the lines of code where the main logic for grouping operations is implemented:

https://github.com/Lightning-AI/lightning-thunder/blob/79e59d0c5c5f8aa8ef80eb31f3fe918466d64c1c/thunder/executors/data_dependent_partition.py#L299-L303

We need to resolve this issue because it leads to memory inefficiency in the generated code with activation checkpointing applied. The `memory_peak_efficient_func` should be the target function for optimization because the same pattern can be found in activation-checkpointed backward functions.

This problem was discovered with Yan's work on enabling PyTorch-native activation checkpointing in https://github.com/Lightning-AI/lightning-thunder/pull/1261. Here are last backward traces https://gist.github.com/kiya00/3ae4890e1ae5abf442d475cccadaa9ec#file-ckp_longchat-7b-16k_traces-py-L2116-L2117.

**Comment by IvanYashchuk** [comment]
@kiya00, @riccardofelluga: could you please work together on resolving this issue? Let me know if you need any support. Thanks!

**Comment by kiya00** [comment]
Hi @riccardofelluga I see you self-assigned, do you think it could be split up, and I can work on a piece, or anything I can help?

**Comment by riccardofelluga** [comment]
Yes we should split it up! I've just assigned it so I can keep track of it

**Comment by kiya00** [comment]
Hi, I noticed the current last_trace of the `memory_peak_efficient_func` is:
```
10 input tensors # t0-t9
10 outputs = nvfusion0(t0, ..., t9) # has 10 Relu fused together
10 matmul take the previous 10 outputs one by one
return [ret0, ..., ret9]
```
if we want memory efficiency, in this case it means like:
```
10 input tensors # t0-t9
out0=nvfusion0(t0) # only 1 ReLU is fused
ret0 = matmul(out0); del out0
out1=nvfusion1(t1) # only 1 ReLU is fused
ret1 = matmul(out1); del out1
... # repeat for 10 inputs
return [ret0, ..., ret9]
```
But that will cause the fusion part to split, am I understanding it right?

**Comment by IvanYashchuk** [comment]
Yes, that's the basic idea. The fusions should be split into individual ones if the computations inside are independent of each other.

**Comment by kiya00** [comment]
I think the easiest way could be to implement a "vertical" merge, otherwise we probably need to investigate if there are other merge algorithms that can balance the memory efficiency and the merge range.

**Comment by IvanYashchuk** [comment]
> when I tried on the longchat-7b-16k, although the sequence of cudnn_sdpa_fwd get separated, the peak memory is even larger, this is the last backward trace: https://gist.github.com/kiya00/4d906a424e0adc7640d3011a0375b5cb

There are quite many operations between the time when [`t1140` and `t1141`](https://gist.github.com/kiya00/4d906a424e0adc7640d3011a0375b5cb#file-longchat-7b-16k_bwd_dfs_topo-txt-L239) from the recomputed forward piece are [actually used](https://gist.github.com/kiya00/4d906a424e0adc7640d3011a0375b5cb#file-longchat-7b-16k_bwd_dfs_topo-txt-L512). Is there anything in terms of dataflow preventing these tensors be placed closer to the backward consumer?

**Comment by kiya00** [comment]
In the [actually used](https://gist.github.com/kiya00/4d906a424e0adc7640d3011a0375b5cb#file-longchat-7b-16k_bwd_dfs_topo-txt-L512) , t1140 and t1141 are hold to wait for t1165 is ready, there's a lone path from [t583](https://gist.github.com/kiya00/4d906a424e0adc7640d3011a0375b5cb#file-longchat-7b-16k_bwd_dfs_topo-txt-L253) until [t1165](https://gist.github.com/kiya00/4d906a424e0adc7640d3011a0375b5cb#file-longchat-7b-16k_bwd_dfs_topo-txt-L508) is calculated.
these 3 tensors need to wait each other somehow, I can try maybe let the smaller tensor wait for the larger tensor to be calculated?

**Comment by kiya00** [comment]
I tried a bottom-up dfs and meanwhile choose the smaller output tensor first:
[t1141](https://gist.github.com/kiya00/c5c05bf24a1462d04eae8340d9cb811d#file-bottom_up_smalltensor_first-txt-L480), [t1140](https://gist.github.com/kiya00/c5c05bf24a1462d04eae8340d9cb811d#file-bottom_up_smalltensor_first-txt-L483), they are [used](https://gist.github.com/kiya00/c5c05bf24a1462d04eae8340d9cb811d#file-bottom_up_smalltensor_first-txt-L487), [t1165](https://gist.github.com/kiya00/c5c05bf24a1462d04eae8340d9cb811d#file-bottom_up_smalltensor_first-txt-L378) is hold to wait for calculating t1140 t1141, but the memory doesn't change much. here are some results, maybe breaking the fusion region causes more memory usage:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/wayan/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/wayan/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


  | dfs experiment | w/ horizontal merge(original)
-- | -- | --
vicuna-33b-v1.3 | 12.29 GB | 12.29 GB
falcon-40b | 19.98 GB | 19.98 GB
Gemma-7b | 21.29 GB | 21.31 GB
longchat-7b-16k | 12.45 GB | 12.36 GB
pythia-12b | 9.65 GB | 9.65 GB



</body>

</html>

**Comment by mruberry** [comment]
@kiya00 @jjsjann123 @IvanYashchuk Would it be possible to let a practitioner or developer control the algorithm used with a developer option? Like maybe the developer option can accept a function whose signature is not guaranteed (but defined somewhere). That could facilitate exploration. We could also think about adding different variants of the algorithm to the repository (maybe in an "experimental" directory) or to a different repository if people want to try them or tweak them.

**Comment by IvanYashchuk** [comment]
Yes, it will be possible to control that with an option.

**Comment by kiya00** [comment]
I tried to add a `memory_efficient_sorting` for our `toposort_bsym_dag`, the jitted function of `memory_peak_efficient/inefficient_func` now uses 0.75MB, but when I tried it on `longchat-7b-16k`, the peak memory is only slightly reduced to 12.19GB, compared to the original 12.45GB

**Comment by kiya00** [comment]
>Awesome! What's the algorithm?

I tried to [set the selector](https://github.com/Lightning-AI/lightning-thunder/compare/main...trytofix1337#diff-e7cd746c8c999010f4b91487688f46f9692ebc2ede34a0f8146eca1955be0eadR289) in `toposort_bsym_dag`, bottom up, always choose the bsym that allocates the maximum memory in the topologically equal node list. Assume the output is allocated by the current bsym, and if the input has no other consumer than the current bsym, it is freed. 

> What's the expected target? Can it be lower than 12.19GB?

Based on the converter branch, before applying the changes in branch(trytofix1337), the memory usage of  longchat-7b-16k(`python thunder/benchmarks/benchmark_litgpt.py --model_name longchat-7b-16k --compile dynamo+thunder --n_layers=2 --max_iters=2 --warmup_iters=1 --checkpoint_activations=True`) is 12.36GB, after is 12.19GB
Inductor costs 12.09 GB, eager costs 10.55GB

**Comment by IvanYashchuk** [comment]
@kiya00 noticed that torch.compile+FSDP+activation checkpointing doesn't hit this fusion and execution order problem because the graph is broken into smaller pieces by Dynamo. https://github.com/Lightning-AI/lightning-thunder/pull/1370 should be enough to enable benchmarks to work in the multi-GPU case.

**Comment by t-vi** [comment]
So one part seems to be that we could refine the creation of the recomputation bsyms. I'm looking into this.
